# Customer Churn Prediction

## Milestone 2: Dataset Collection & Data Understanding

### Business Problem

Customer churn is one of the biggest challenges faced by telecom companies. Acquiring a new customer is often more expensive than retaining an existing one. Therefore, predicting customers who are likely to leave the service can help businesses take proactive measures to improve customer retention.

### Objective

The objective of this notebook is to understand the dataset before performing any preprocessing or machine learning. We will inspect the dataset, identify missing values, analyze data types, and make initial observations that will guide the next stages of the project.

In [1]:
# Import necessary libraries

import pandas as pd
import numpy as np

# Display all columns
pd.set_option("display.max_columns", None)

# Display numbers with 2 decimal places
pd.set_option("display.float_format", "{:.2f}".format)

## Loading the Dataset

In this section, we load the IBM Telco Customer Churn dataset into a Pandas DataFrame. This allows us to inspect and analyze the data efficiently.

In [2]:
# Load the dataset

df = pd.read_csv("../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv")

# Display first five rows
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## Dataset Overview

Before performing any analysis, it is important to understand the overall size of the dataset. We will check the number of rows and columns.

In [3]:
# Check dataset dimensions

print(f"Number of Rows    : {df.shape[0]}")
print(f"Number of Columns : {df.shape[1]}")

Number of Rows    : 7043
Number of Columns : 21


## Dataset Information

The `info()` method provides details about each column, including the data type, number of non-null values, and memory usage. This helps us identify missing values and understand the structure of the dataset.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


## Displaying Sample Records

To better understand the data, we inspect a few records from both the beginning and the end of the dataset.

In [5]:
# Display first five rows

df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [6]:
# Display last five rows

df.tail()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,No,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,Yes,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,No,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes
7042,3186-AJIEK,Male,0,No,No,66,Yes,No,Fiber optic,Yes,No,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),105.65,6844.5,No


In [7]:
# Generate descriptive statistics

df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.00,7043.00,7043.00
mean,0.16,32.37,64.76
std,0.37,24.56,30.09
min,0.00,0.00,18.25
25%,0.00,9.00,35.50
50%,0.00,29.00,70.35
75%,0.00,55.00,89.85
max,1.00,72.00,118.75


In [8]:
# Descriptive statistics including categorical columns

df.describe(include="object")

,customerID,gender,Partner,Dependents,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,TotalCharges,Churn
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043
unique,7043,2,2,2,2,3,3,3,3,3,3,3,3,3,2,4,6531,2
top,7590-VHVEG,Male,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,,No
freq,1,3555,3641,4933,6361,3390,3096,3498,3088,3095,3473,2810,2785,3875,4171,2365,11,5174


## Missing Value Analysis

Before preprocessing the dataset, we need to check whether any columns contain missing values.

Missing values can negatively affect machine learning models and often require cleaning or imputation.

In [9]:
# Check missing values

missing_values = df.isnull().sum()

missing_values

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [10]:
# Display only columns with missing values

missing_values[missing_values > 0]

Series([], dtype: int64)

## Duplicate Record Analysis

Duplicate records can introduce bias into machine learning models.

Here, we verify whether the dataset contains any duplicate rows.

In [11]:
# Count duplicate rows

df.duplicated().sum()

np.int64(0)

## Unique Values in Each Column

Understanding the number of unique values in each feature helps identify categorical variables, high-cardinality columns, and potential identifiers.

In [12]:
# Count unique values in each column

df.nunique()

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64

## Target Variable Analysis

The target variable for this project is **Churn**.

We inspect its unique values and class distribution before model training.

In [13]:
# Unique values in target column

df["Churn"].unique()

array(['No', 'Yes'], dtype=object)

In [14]:
# Count customers in each class

df["Churn"].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [15]:
# Percentage distribution

df["Churn"].value_counts(normalize=True) * 100

Churn
No    73.46
Yes   26.54
Name: proportion, dtype: float64

## Checking the Data Type of TotalCharges

Although `TotalCharges` represents a numeric value, it was loaded as an object data type.

We investigate why this happened.

In [16]:
# Display data type

df["TotalCharges"].dtype

dtype('O')

In [17]:
# Check for blank spaces

(df["TotalCharges"] == " ").sum()

np.int64(11)

## Initial Observations

- The dataset contains 7043 customer records.
- There are 21 columns.
- No duplicate records were found.
- CustomerID is a unique identifier and will not be useful for prediction.
- The target variable is imbalanced, with approximately 73% non-churn customers and 27% churn customers.
- The `TotalCharges` column has an incorrect data type due to blank string values.
- Data preprocessing will include handling these blank values, converting data types, encoding categorical variables, and feature scaling.

## Data Dictionary

| Column | Description |
|--------|-------------|
| customerID | Unique identifier assigned to each customer |
| gender | Customer's gender |
| SeniorCitizen | Indicates whether the customer is a senior citizen (1 = Yes, 0 = No) |
| Partner | Whether the customer has a partner |
| Dependents | Whether the customer has dependents |
| tenure | Number of months the customer has stayed with the company |
| PhoneService | Whether the customer has phone service |
| MultipleLines | Whether the customer has multiple phone lines |
| InternetService | Type of internet service subscribed |
| OnlineSecurity | Whether online security service is subscribed |
| OnlineBackup | Whether online backup service is subscribed |
| DeviceProtection | Whether device protection service is subscribed |
| TechSupport | Whether technical support service is subscribed |
| StreamingTV | Whether streaming TV service is subscribed |
| StreamingMovies | Whether streaming movie service is subscribed |
| Contract | Customer's contract type |
| PaperlessBilling | Whether paperless billing is enabled |
| PaymentMethod | Customer's payment method |
| MonthlyCharges | Monthly amount charged to the customer |
| TotalCharges | Total amount charged to the customer |
| Churn | Target variable indicating whether the customer left the service |

# Conclusion

The IBM Telco Customer Churn dataset contains 7,043 customer records and 21 features. Initial inspection revealed no duplicate records and no standard missing values. However, the `TotalCharges` column contains 11 blank string values, which caused it to be stored as an object instead of a numeric data type.

The target variable (`Churn`) is moderately imbalanced, with approximately 73.46% non-churn customers and 26.54% churn customers. This class imbalance will be addressed during preprocessing using SMOTE after the train-test split.

Overall, the dataset is well-structured and suitable for building a customer churn prediction model after appropriate preprocessing.